risc0/zkvm/src/host/api/server.rs

```diff
@@ -323,39 +324,89 @@ impl Server {
             let bytes = binary.as_bytes()?;
             let mut exec = ExecutorImpl::from_elf(env, &bytes)?;
 
-            let session = exec.run_with_callback(|segment| {
-                let segment_bytes = bincode::serialize(&segment)?;
-                let asset = pb::api::Asset::from_bytes(
-                    &segments_out,
-                    segment_bytes.into(),
-                    format!("segment-{}", segment.index),
-                )?;
-                let msg = pb::api::ServerReply {
-                    kind: Some(pb::api::server_reply::Kind::Ok(pb::api::ClientCallback {
-                        kind: Some(pb::api::client_callback::Kind::SegmentDone(
-                            pb::api::OnSegmentDone {
-                                segment: Some(pb::api::SegmentInfo {
-                                    index: segment.index,
-                                    po2: segment.inner.po2 as u32,
-                                    cycles: segment.inner.insn_cycles as u32,
-                                    segment: Some(asset),
-                                }),
-                            },
-                        )),
-                    })),
-                };
-                tracing::trace!("tx: {msg:?}");
-                conn.send(msg)?;
-
-                let reply: pb::api::GenericReply = conn.recv()?;
-                tracing::trace!("rx: {reply:?}");
-                let kind = reply.kind.ok_or(malformed_err())?;
-                if let pb::api::generic_reply::Kind::Error(err) = kind {
-                    bail!(err)
-                }
-
-                Ok(Box::new(NullSegmentRef))
-            })?;
+            // use redis if available
+            let session = if let AssetRequest::Redis(url, key, ttl) =
+                AssetRequest::try_from(segments_out.clone())?
+            {
+                let client = redis::Client::open(url)?;
+                let mut connection = client.get_connection()?;
+                exec.run_with_callback(|segment| {
```

Comment:
Could we use a chain of responsibility pattern for this?

Reply:
Could I bug you for an example/pseudocode? I'm aware of the pattern, just not sure how to approach it for this specific use case.